core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -330,18 +330,45 @@ object SparkSubmit extends CommandLineUtils {
     args.archives = Option(args.archives).map(resolveGlobPaths(_, hadoopConf)).orNull
 
     // In client mode, download remote files.
+    var localPrimaryResource: String = null
+    var localJars: String = null
+    var localPyFiles: String = null
     if (deployMode == CLIENT) {
-      args.primaryResource = Option(args.primaryResource).map {
+      localPrimaryResource = Option(args.primaryResource).map {
         downloadFile(_, targetDir, args.sparkProperties, hadoopConf)
       }.orNull
-      args.jars = Option(args.jars).map {
+      localJars = Option(args.jars).map {
         downloadFileList(_, targetDir, args.sparkProperties, hadoopConf)
       }.orNull
-      args.pyFiles = Option(args.pyFiles).map {
+      localPyFiles = Option(args.pyFiles).map {
         downloadFileList(_, targetDir, args.sparkProperties, hadoopConf)
       }.orNull
     }
 
+    if (clusterManager == YARN) {
+      def isNoneFsFileExist(paths: String): Boolean = {
+        Option(paths).exists { p =>
+          p.split(",").map(_.trim).filter(_.nonEmpty).exists { path =>
+            val url = Utils.resolveURI(path)
+            url.getScheme match {
+              case "http" | "https" | "ftp" => true
+              case _ => false
+            }
+          }
+        }
+      }
+
+      // Spark on YARN doesn't support upload remote resources from http, https or ftp server
+      // directly to distributed cache, so print a warning and exit the process.
+      if (isNoneFsFileExist(args.jars) ||
```

Comment:
I also think this is a separate issue. It didn't work before, it's ok if it doesn't work now. Maybe it works in 2.2, but I think it's safe to assume people aren't really depending on that since 2.2 is very recent.

This change should be about avoiding to re-upload files downloaded by the client, and we can look at fixing http-on-YARN in a different patch.